scraper/scraper/spiders/meps.py

```diff
@@ -0,0 +1,68 @@
+from scrapy.spiders import XMLFeedSpider
+from scraper.items import MEP
+import scrapy
+
+import os
+
+
+class XMLParser(XMLFeedSpider):
+    name = 'mep_xml'
+    start_urls = ['https://www.europarl.europa.eu/meps/en/directory/xml']
+    iterator = 'iternodes'
+    itertag = 'mep'
+
+    def parse_node(self, response, node):
+        id = node.xpath('id/text()').get()
+        full_name = node.xpath('fullName/text()').get()
+
+        return MEP(id=id, full_name=full_name)
+
+
+class MEPSCrawler(scrapy.Spider):
+    name = 'mep'
+    base_url = 'https://www.europarl.europa.eu'
+    has_declaration = 0
+
+    def start_requests(self):
+        if os.path.isdir(self.dir_name) is False:
+            os.mkdir(self.dir_name)
+        yield scrapy.Request(url=self.url, callback=self.parse_table)
+
+    def parse_table(self, response):
+        # get the table of contents on the left side and extract all links
+        table = response.xpath(('//aside[@class="ep_gridcolumn '
+                                'ep-layout_tableofcontent"]'))
+        links = [self.base_url + x for x in table.xpath('.//a/@href').getall()
+                 if x.startswith('/meps/')]
+
+        for link in links:
+            yield scrapy.Request(url=link, callback=self.parse_page,
+                                 dont_filter=True)
+
+    def parse_page(self, response):
+        page_content = response.xpath('//div[@class="ep_gridcolumn"]')
+        title = page_content.xpath('.//span[@class="ep_name"]//text()').get()
+
+        with open(f'{self.dir_name}/{title}.html', 'w') as f:
+            f.write(page_content.get())
+
+        declaration = page_content.xpath('.//span[@class="ep_name" '
+                                         'and text()= "Declarations"]')
+
+        # if in declaration section download all files
+        if declaration.get():
+            self.has_declaration = 1
+            docs_section = declaration.xpath(('.//..//..//..//..//..//..//..//'
```

Comment:
`declaration` looks for a `span` that contains the text 'Declarations'. That `span` is usually located just above the pdfs that we want, but it is buried very deep, in 8 levels of `div`. This xpath just gets to the appropriate level and moves to the `div` that contains our lovely pdfs.

Reply:
I see it now. It would be safer to declare what those `..` elements are: `.//div//div//aside//nav//ol//li//div//span//span//`.

You have multiple top-level `//div[@class="ep_gridrow"]` so it makes sense to start from those instead of `ep_gridcolumn`. You should also get the `declaration` from that long xpath above, instead of just taking the first `span.ep_name` with matching text.

The target here is to make the scraper crash and die on the first trace of a page structure change.